framework/di/Container.php

```diff
@@ -265,16 +265,21 @@ public function set($class, $definition = [], array $params = [])
      *
      * @param string $class class name, interface name or alias name
      * @param mixed $definition the definition associated with `$class`. See [[set()]] for more details.
+     * If `null`, next call to [[get()]] behaves in the same way as if the instance was never created.
      * @param array $params the list of constructor parameters. The parameters will be passed to the class
      * constructor when [[get()]] is called.
      * @return $this the container itself
      * @see set()
      */
     public function setSingleton($class, $definition = [], array $params = [])
     {
-        $this->_definitions[$class] = $this->normalizeDefinition($class, $definition);
-        $this->_params[$class] = $params;
-        $this->_singletons[$class] = null;
+        if ($definition !== null) {
+            $this->_definitions[$class] = $this->normalizeDefinition($class, $definition);
+            $this->_params[$class] = $params;
+            $this->_singletons[$class] = null;
+        } elseif (array_key_exists($class, $this->_singletons)) {
+            $this->_singletons[$class] = null;
```

Comment:
as far as I see there is no need for the elseif and `$this->_singletons[$class] = null;` can be performed in any case.